File: api/controllers/auth.controller.go
package controllers

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"net/http"
	"os"
	"time"

	"github.com/dgrijalva/jwt-go"
	"github.com/gin-gonic/gin"
	"golang.org/x/crypto/bcrypt"
	"go.mongodb.org/mongo-driver/bson"
	
	"gopkg.in/gomail.v2"

	"api/config"
	
)

// User represents a simple user model.
type User struct {
	ID           interface{} `bson:"_id,omitempty" json:"id,omitempty"`
	Username     string      `bson:"username" json:"username"`
	Email        string      `bson:"email" json:"email"`
	Password     string      `bson:"password" json:"password"`
	Avatar       string      `bson:"avatar,omitempty" json:"avatar,omitempty"`
	ResetToken   string      `bson:"reset_token,omitempty" json:"-"`
	ResetExpires int64       `bson:"reset_expires,omitempty" json:"-"`
	CreatedAt    time.Time   `bson:"createdAt" json:"createdAt"`
}

// LoginInput represents the expected login payload.
type LoginInput struct {
	Email    string `json:"email" binding:"required,email"`
	Password string `json:"password" binding:"required"`
}

// JWTClaims defines the structure for JWT claims.
type JWTClaims struct {
	ID string `json:"id"`
	jwt.StandardClaims
}

// generateToken creates a JWT token for the given userID.
func generateToken(userID string) (string, error) {
	claims := JWTClaims{
		ID: userID,
		StandardClaims: jwt.StandardClaims{
			ExpiresAt: time.Now().Add(24 * time.Hour).Unix(),
		},
	}
	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	return token.SignedString([]byte(os.Getenv("JWT_SECRET")))
}

// hashPassword hashes the provided password.
func hashPassword(password string) (string, error) {
	bytes, err := bcrypt.GenerateFromPassword([]byte(password), 10)
	return string(bytes), err
}

// verifyPassword compares a hashed password with its plaintext equivalent.
func verifyPassword(hashed, plain string) bool {
	err := bcrypt.CompareHashAndPassword([]byte(hashed), []byte(plain))
	return err == nil
}

// Signup registers a new user.
func Signup(c *gin.Context) {
	var user User

	// Bind incoming JSON to the user struct.
	if err := c.ShouldBindJSON(&user); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid input"})
		return
	}

	// Hash the password.
	hashedPassword, err := hashPassword(user.Password)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to hash password"})
		return
	}
	user.Password = hashedPassword
	user.CreatedAt = time.Now()

	collection := config.DB.Collection("users")
	_, err = collection.InsertOne(context.TODO(), user)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "User creation failed"})
		return
	}

	c.JSON(http.StatusCreated, gin.H{"success": true, "message": "User registered successfully"})
}

// Signin logs in a user.
func Signin(c *gin.Context) {
	var input LoginInput
	if err := c.ShouldBindJSON(&input); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid input"})
		return
	}

	// Get the users collection from the MongoDB database.
	collection := config.DB.Collection("users")

	var user User
	// Look up the user by email.
	err := collection.FindOne(context.TODO(), bson.M{"email": input.Email}).Decode(&user)
	if err != nil || !verifyPassword(user.Password, input.Password) {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "Invalid credentials"})
		return
	}

	// Convert user.ID to a string for token generation.
	// Depending on your implementation, user.ID may be of type primitive.ObjectID.
	userIDStr := fmt.Sprintf("%v", user.ID)
	token, err := generateToken(userIDStr)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to generate token"})
		return
	}

	// Optionally, you might want to omit the password from the returned user object.
	user.Password = ""

	c.JSON(http.StatusOK, gin.H{"token": token, "user": user})
}
// ForgotPassword sends a password reset link.
func ForgotPassword(c *gin.Context) {
	var input struct{ Email string }
	if err := c.ShouldBindJSON(&input); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid email"})
		return
	}

	var user User
	collection := config.DB.Collection("users")
	err := collection.FindOne(context.TODO(), bson.M{"email": input.Email}).Decode(&user)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "User not found"})
		return
	}

	tokenBytes := make([]byte, 32)
	_, err = rand.Read(tokenBytes)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to generate token"})
		return
	}
	resetToken := hex.EncodeToString(tokenBytes)

	update := bson.M{"$set": bson.M{
		"reset_token":   resetToken,
		"reset_expires": time.Now().Add(time.Hour).Unix(),
	}}
	_, err = collection.UpdateOne(context.TODO(), bson.M{"email": input.Email}, update)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update reset token"})
		return
	}

	resetLink := os.Getenv("FRONTEND_URL") + "/reset-password/" + resetToken
	sendResetEmail(user.Email, resetLink)

	c.JSON(http.StatusOK, gin.H{"message": "Password reset link sent!"})
}

// ResetPassword resets the user's password.
func ResetPassword(c *gin.Context) {
	var input struct {
		Token    string `json:"token"`
		Password string `json:"password"`
	}
	if err := c.ShouldBindJSON(&input); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid input"})
		return
	}

	var user User
	collection := config.DB.Collection("users")
	err := collection.FindOne(context.TODO(), bson.M{
		"reset_token":   input.Token,
		"reset_expires": bson.M{"$gt": time.Now().Unix()},
	}).Decode(&user)
	if err != nil {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "Invalid or expired token"})
		return
	}

	hashedPassword, err := hashPassword(input.Password)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to hash password"})
		return
	}
	update := bson.M{"$set": bson.M{"password": hashedPassword, "reset_token": nil, "reset_expires": nil}}
	_, err = collection.UpdateOne(context.TODO(), bson.M{"email": user.Email}, update)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to reset password"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Password has been reset"})
}

// sendResetEmail sends a password reset email.
func sendResetEmail(email, link string) {
	m := gomail.NewMessage()
	m.SetHeader("From", os.Getenv("EMAIL_USER"))
	m.SetHeader("To", email)
	m.SetHeader("Subject", "Password Reset Request")
	m.SetBody("text/html", "Click <a href='"+link+"'>here</a> to reset your password.")

	d := gomail.NewDialer("smtp.gmail.com", 587, os.Getenv("EMAIL_USER"), os.Getenv("EMAIL_PASS"))
	d.DialAndSend(m)
}

File: api/controllers/auth.controller.js
import User from "../models/user.model.js";
import bcryptjs from "bcryptjs";
import { errorHandler } from "../utils/error.js";
import jwt from "jsonwebtoken";
import { validationResult } from "express-validator";
import crypto from "crypto";
import nodemailer from "nodemailer";

// API for admin login
export const loginAdmin = async (req, res) => {
  try {
    const { email, password } = req.body;
    if (
      email === process.env.ADMIN_EMAIL &&
      password === process.env.ADMIN_PASSWORD
    ) {
      const token = jwt.sign(email + password, process.env.JWT_SECRET);
      res.json({ success: true, token });
    } else {
      res.json({ success: false, message: "Invalid credentials" });
    }
  } catch (error) {
    console.log(error);
    res.json({ success: false, message: error.message });
  }
};

// Register
export const signup = async (req, res, next) => {
  const errors = validationResult(req);
  if (!errors.isEmpty()) {
    return res.status(400).json({ status: 400, errors: errors.array() });
  }
  const { username, email, password } = req.body;
  try {
    const hashedPassword = bcryptjs.hashSync(password, 10);
    const newUser = new User({ username, email, password: hashedPassword });
    await newUser.save();
    res.status(201).json({
      status: 201,
      message: "User created successfully!",
      user: { username: newUser.username, email: newUser.email },
    });
  } catch (error) {
    next(error);
  }
};

// Login
export const signin = async (req, res, next) => {
  try {
    const { email, password } = req.body;
    const validUser = await User.findOne({ email });
    if (!validUser || !bcryptjs.compareSync(password, validUser.password)) {
      return next(errorHandler(401, "Wrong credentials!"));
    }
    const token = jwt.sign({ id: validUser._id }, process.env.JWT_SECRET);
    const { password: pass, ...rest } = validUser._doc;

    res.status(200).json({ rest, token });
  } catch (error) {
    next(error);
  }
};

// Login With Google
export const google = async (req, res, next) => {
  try {
    const user = await User.findOne({ email: req.body.email });
    if (user) {
      const token = jwt.sign({ id: user._id }, process.env.JWT_SECRET);
      const { password: pass, ...rest } = user._doc;
      res.status(200).json({ rest, token });
    } else {
      const generatedPassword =
        Math.random().toString(36).slice(-8) +
        Math.random().toString(36).slice(-8);
      const hashedPassword = bcryptjs.hashSync(generatedPassword, 10);
      const newUser = new User({
        username:
          req.body.name.split(" ").join("").toLowerCase() +
          Math.random().toString(36).slice(-4),
        email: req.body.email,
        password: hashedPassword,
        avatar: req.body.photo,
      });
      await newUser.save();
      const token = jwt.sign({ id: newUser._id }, process.env.JWT_SECRET);
      const { password: pass, ...rest } = newUser._doc;
      res.status(200).json({ rest, token });
    }
  } catch (error) {
    next(error);
  }
};

// Logout
export const signOut = async (req, res, next) => {
  try {
    res.status(200).json({ status: 200, message: "User has been logged out!" });
  } catch (error) {
    next(error);
  }
};

// Forgot Password Request
export const forgotPassword = async (req, res, next) => {
  try {
    const { email } = req.body;
    const user = await User.findOne({ email });

    if (!user) {
      return next(errorHandler(404, "User with this email does not exist!"));
    }
    const resetToken = crypto.randomBytes(32).toString("hex");
    const hashedToken = bcryptjs.hashSync(resetToken, 10);
    user.resetPasswordToken = hashedToken;
    user.resetPasswordExpires = Date.now() + 3600000;

    await user.save();

    const resetLink = `${process.env.FRONTEND_URL}/reset-password/${resetToken}`;
    const transporter = nodemailer.createTransport({
      service: "Gmail",
      auth: {
        user: process.env.EMAIL_USER,
        pass: process.env.EMAIL_PASS,
      },
    });

    const mailOptions = {
      to: user.email,
      from: process.env.EMAIL_USER,
      subject: "Password Reset Request",
      html: `<p>You are receiving this email because you (or someone else) requested to reset the password for your account.</p>
             <p>Please click on the following link, or paste this into your browser to complete the process:</p>
             <p><a href="${resetLink}">${resetLink}</a></p>
             <p>If you did not request this, please ignore this email and your password will remain unchanged.</p>`,
    };

    await transporter.sendMail(mailOptions);

    res.status(200).json({
      status: 200,
      message: "Password reset link sent to your email!",
    });
  } catch (error) {
    next(error);
  }
};

// Reset Password
export const resetPassword = async (req, res, next) => {
  try {
    const { token, password } = req.body;
    const user = await User.findOne({
      resetPasswordExpires: { $gt: Date.now() },
    });

    if (!user || !bcryptjs.compareSync(token, user.resetPasswordToken)) {
      return next(errorHandler(400, "Invalid or expired token!"));
    }

    const hashedPassword = bcryptjs.hashSync(password, 10);
    user.password = hashedPassword;
    user.resetPasswordToken = undefined;
    user.resetPasswordExpires = undefined;

    await user.save();

    res.status(200).json({
      status: 200,
      message: "Password has been reset successfully!",
    });
  } catch (error) {
    next(error);
  }
};

// Validate Reset Token
export const validateResetToken = async (req, res, next) => {
  try {
    const { token } = req.params;

    const user = await User.findOne({
      resetPasswordExpires: { $gt: Date.now() },
    });

    if (!user || !bcryptjs.compareSync(token, user.resetPasswordToken)) {
      return next(errorHandler(400, "Invalid or expired token!"));
    }

    res.status(200).json({
      status: 200,
      message: "Reset token is valid.",
    });
  } catch (error) {
    next(error);
  }
};
